lib/resty/mlcache.lua

```diff
@@ -670,7 +670,31 @@ function _M:get(key, opts, cb, ...)
 
     local elapsed, err = lock:lock(LOCK_KEY_PREFIX .. namespaced_key)
     if not elapsed then
-        return nil, "could not aquire callback lock: " .. err
+        if err ~= "timeout" then
+            return nil, "could not aquire callback lock: " .. err
+        end
+
+        -- in case of a timeout, there's still hope: if the initial worker
+        -- that held the lock released it and then another locked it to see if
+        -- the result was arrived at the wrong time, this worker might timeout
+        -- because of that.
+        -- This event has a hight probability to happen when the steps in the
+        -- spin-lock loop get bigger.
+        data, err = get_shm_set_lru(self, key, namespaced_key)
+        if err then
+            return nil, err
+        end
+
+        if data == CACHE_MISS_SENTINEL_LRU then
+            return nil, nil, 2
+        end
+
+        if data ~= nil then
+            return data, nil, 2
+        end
+
+        -- ok, this time there's no more hope
+        return nil, "could not aquire callback lock: timeout"
```

Comment:
@jdesgats This is a great catch, thank you!

It seems to me like the following patch would achieve the same behaviour in a less invasive way: 
```diff
diff --git a/lib/resty/mlcache.lua b/lib/resty/mlcache.lua
index 35b0db18..ec3ae039 100644
--- a/lib/resty/mlcache.lua
+++ b/lib/resty/mlcache.lua
@@ -607,7 +607,7 @@ end
 
 local function unlock_and_ret(lock, res, err, hit_lvl)
     local ok, lerr = lock:unlock()
-    if not ok then
+    if not ok and lerr ~= "unlocked" then
         return nil, "could not unlock callback: " .. lerr
     end
 
@@ -668,9 +668,9 @@ function _M:get(key, opts, cb, ...)
         return nil, "could not create lock: " .. err
     end
 
-    local elapsed, err = lock:lock(LOCK_KEY_PREFIX .. namespaced_key)
-    if not elapsed then
-        return nil, "could not aquire callback lock: " .. err
+    local elapsed, lerr = lock:lock(LOCK_KEY_PREFIX .. namespaced_key)
+    if not elapsed and lerr ~= "timeout" then
+        return nil, "could not aquire callback lock: " .. lerr
     end
 
     -- check for another worker's success at running the callback
@@ -680,14 +680,22 @@ function _M:get(key, opts, cb, ...)
         return unlock_and_ret(lock, nil, err)
     end
 
-    if data then
-        if data == CACHE_MISS_SENTINEL_LRU then
-            return unlock_and_ret(lock, nil, nil, 2)
-        end
+    if data == CACHE_MISS_SENTINEL_LRU then
+        return unlock_and_ret(lock, nil, nil, 2)
+    end
 
+    if data ~= nil then
         return unlock_and_ret(lock, data, nil, 2)
     end
 
+    -- data is nil, we are either the 1st worker to hold the lock, or
+    -- a subsequent worker whose lock has timed out before the 1st one
+    -- finished to run the callback.
+
+    if lerr == "timeout" then
+        return nil, "could not acquire callback lock: timeout"
+    end
+
     -- still not in shm, we are responsible for running the callback
 
     local pok, perr, err, new_ttl = pcall(cb, ...)
```

What do you think?

Reply:
Yeah it looks that you approach works as well in a cleaner way without adding even more branches to the already long `:get()` method :+1: 

Feel free to decline my PR and commit your fix.

Reply:
It is fine with me if you want to replace your patch with this one as well. If not, I will commit this patch tomorrow and credit you in the commit :+1: